### DISCUSSION TITLE
Spindle sync with blackpill board

### DISCUSSION THREAD

**Discussion by jal123sin** [grblHAL/core#354] [Q&A]
Hi, I am using the MKS DLC32 controller for my lathe now. I want to use the spindle sync mode but it is not available to select spindle sync mode on the Web Builder. Therefore, I am considering to switch to the blackpill series but I cannot find the rpm counter pin on the stm32F411 board map. Now my spindle driver have the 0-5v pwm and please advise do I need the real time rpm check for the blackpill.
Or, do you consider to add spindle sync mode on the board MKS DLC32 in near future, thanks.

**Comment by terjeio** [comment]
> Therefore, I am considering to switch to the blackpill series but I cannot find the rpm counter pin on the stm32F411 board map.

It is defined [here](https://github.com/grblHAL/STM32F4xx/blob/0d20911eda43b7a31ba28b597d1d6968eb6c0e87/Inc/blackpill_map.h#L119-L129).

> Or, do you consider to add spindle sync mode on the board MKS DLC32 in near future, thanks.

No, not by me since the ESP32 FPU crashes the controller when use in an interrupt context.
The PID control code has to be extended to support fixed point math, driver extended and retested - if somebody is willing to do that and provide a PR then it may eventually happen.

**Comment by jal123sin** [comment]
Thanks for quick response, I am going to use the hall sensor, in which it have 3pins (5v, ground and signal). Could you please advise the sensor signal pin connect to either pin B3 or pin A15.
![Pic1](https://github.com/grblHAL/core/assets/136598254/1675a89c-aa3a-4ad0-a5ad-ac8535243850)

**Reply by terjeio** [reply]
You need two sensors, one for the index (one pulse per revolution) and one for the pulses (> 1 pulse per revolution, I have 120 on my lathe).
I have not tried with connecting the two to the same sensor and don't know if that would work.
If the sensor is only for RPM feedback (not spindle synced motion) then connect the sensor to the pulse pin.

**Reply by duyna22** [reply]
Hello, 

"If the sensor is only for RPM feedback (not spindle synced motion) then connect the sensor to the pulse pin."
- That's mean we only need pin A15 to receive RPM from sensor? Pin B3 will be leaved idle (no connection) ?
- Do we need to configure anything else?  (Setting $38)

Thank you.

**Reply by terjeio** [reply]
Setting $38 has to be set to number of pulses per spindle revolution, it is used for the RPM calculation. The index pulse is not needed for that.

**Reply by duyna22** [reply]
So, summary: 

We can enable Spindle Sync by sending RPM signal to Pin A15 and setting $38 = number of pulses per spindle revolution, right?

**Reply by terjeio** [reply]
No, for spindle sync you need the index signal as well.

**Reply by duyna22** [reply]
Hello,

Sorry for my knowledge, I really don't know "Spindle Index Signal" and "Spindle Pulse Signal". I search on Google, it's all about Encoder Index, Encoder A, Encoder B.
Can you share tutorial or document about two signal that connect to Pin A15 and B3? The wire connection and detail meaning..

Thank you.

**Reply by terjeio** [reply]
[This video](https://www.youtube.com/watch?v=yY_4HwmZEaQ) helps?

**Reply by duyna22** [reply]
Hi, 

I have AS5600 magnet sensor, 
And I attached it into a Stepper Motor with 1.8degree/step, That's mean I can measure:
- Stepper Motor Pulse
- 200ppr (360/1.8)
- RPM 
- Angle Position
- Steps (total)
- Steps/s

So, With all values above, what values can be use to enable Spindle Sync ?

Thank you.

**Reply by terjeio** [reply]
> I have AS5600 magnet sensor,

You cannot use this since it has an I2C interface, a similar sensor that [provides ABI output](https://ams.com/angle-position-on-axis) should be ok though - only A+I or B+I is used by the spindle sync code.

**Comment by jal123sin** [comment]
I learned from one of the grblhal lathe project and copy the photo as below. You said two sensors and 120, are you meaning that the total disc holes is 120 as shown on the photo. 
![Pic2](https://github.com/grblHAL/core/assets/136598254/e0633cde-5beb-4f26-9afc-0249f3553f93)

**Reply by terjeio** [reply]
> . You said two sensors and 120, are you meaning that the total disc holes is 120 as shown on the photo.

Yes. But number of pulses (holes) can be different - setting $38 is used for configuring number of pulses per revolution (PPR).
More info can be found [in the wiki](https://github.com/grblHAL/core/wiki/Spindle-sync).
